### DISCUSSION TITLE
the CORS header problem

### DISCUSSION THREAD

**Discussion by rew1ndustry** [arendst/Tasmota#21540] [Q&A]
am new to all this, have tried to follow "Local access security issue due to HTTP CORS header #6767", and "Set CORS to be an optional feature #12827", however i now understand that CORS is gone.

i cannot understand why Tasmota includes a very complete http API that seems to be essentially useless, without CORS.

as far as i can tell there is no way to have the device host my pages and, as far as i see this is the most secure answer to the problem, by eliminating the boundary issue.  am i mistaken to believe that Tasmota cannot do this?  if so, why not?

as far as i can tell the API is essentially useless - at least in terms of responsive applications - without proper CORS headers.  what am i not understanding?  is there some other way to get responses from commands over the http API?

as far as i can tell the API continues to honour incoming commands and, although i get no feedback, i have no problem controlling my devices over the http API.  so what security problem have you fixed, by removing CORS?

or, in general, can anyone please help me with what i am not understanding?

i am limited to firefox, running on macos catalina - is there something i don't know about this setup that is corrupting my results?

or is there a discussion i have missed that explains properly why CORS is gone, and how we proceed without?

on a final note, i do understand Tasmota is all about MQTT, but i cannot run brokers and must have point to point control, without anything third party - hence my need for the http API.  is it possible to get Tasmota without the MQTT overhead?  perhaps there is another project that would suit me better?

am sorry for all these questions, i know i am being very muddled - anyone please help?

**Comment by s-hadinger** [comment]
> as far as i can tell the API is essentially useless - at least in terms of responsive applications - without proper CORS headers. what am i not understanding? is there some other way to get responses from commands over the http API?

You make the assumption that APIs are only used by web browser, but this is only one use-case among many. For example, Tasmota Matter uses HTTP API to bridge ESP8266 based devices. Since they are just plain HTTP requests made from a non-browser client, it does not use nor need CORS.

If you want to enable CORS, you can do it but it does causes some potential security issues that you need to be aware of.

**Comment by rew1ndustry** [comment]
thank you, i am embarrassed - i was advised that CORS had been entirely discontinued.

am i correct to understand, from #12827, i would have to download and cross-compile Tasmota, in order to activate CORS support?

if so, frankly, i would sooner put the effort into writing my own firmware.

not to disparage Tasmota in any way - this is important work.

as an alternative, if i can find the time, i could take a look at your CORS issues, if you like?

can do good solid code, would need help understanding exactly what needs to be improved, to make it properly secure - however am happy to learn.
